docs/howto/push-notifications.md

```diff
@@ -170,6 +170,64 @@ idiosyncratic capitalization shown.
 
 ### Current workaround
 
+This workaround means using a development server. If that won't work for you,
+you'll have to try the second workaround, below.
+
+We can tell our development server to talk to Apple's APNs "sandbox" server,
+instead of its server meant for production, but you'll need a certificate signed
+by Apple authorizing you to do so. Some background on that is
+[here](https://developer.apple.com/documentation/usernotifications/setting_up_a_remote_notification_server/sending_notification_requests_to_apns/#2947606).
+
+1. First, generate a Certificate Signing Request (CSR). Apple Developer's
+   instructions are
+   [here](https://help.apple.com/developer-account/#/devbfa00fef7). You can use
+   something like "Johnny Appleseed APNs Sandbox" for the "Common Name."
+
+2. Greg is authorized in Apple Developer to upload the CSR and obtain the actual
+   certificate, so you should send it to him and ask him to do that. He'll
+   follow something like [these
+   instructions](https://developer.apple.com/documentation/usernotifications/setting_up_a_remote_notification_server/establishing_a_certificate-based_connection_to_apns)
+   to upload your CSR and obtain an APNs SSL (Sandbox) certificate (not Sandbox
+   & Production). (Note: When you created the CSR, Keychain Access also created
+   a private key. The CSR itself is not the private key, so it doesn't require
+   extra care to send over the Internet. The APNs certificate doc **wrongly**
```

Comment:
The offending doc is at https://developer.apple.com/documentation/usernotifications/setting_up_a_remote_notification_server/establishing_a_certificate-based_connection_to_apns; Greg pointed out to me that this doc is, indeed, in error — not us.

Still, if there's a good way to do it, I'd like to succinctly allay a contributor's fear that _we're_ the ones who are wrong here. Ideas?

Reply:
Less bold text, for one. Perhaps something like the following?

> Note that, as of 2020-03-02, the linked Apple documentation incorrectly conflates the CSR and the private key. The CSR contains only *<the public key / a hash of the private key>*, which can be confirmed by *<whatever you and/or Greg did to confirm this>*.

Reply:
> <whatever you and/or Greg did to confirm this>

It was from Greg's knowledge of how CSR's work in general.

I believe it was this: 1) everyone who sets up their website to use SSL issues a CSR to a Certificate Authority over the Internet; 2) assume the CSR contains the private key; 3) either people's private keys are intercepted and stolen all the time, or no one issues CSR's because that's a possibility, and 4) contradiction, we're not seeing step 3.

@gnprice But this is a bit verbose; I think even a link to a contradictory doc from Apple would be fine.

Reply:
There's [a relevant bit on Wikipedia](https://en.wikipedia.org/w/index.php?title=Certificate_signing_request&oldid=940627604#Structure) – although that's not very meaningful; generally if Wikipedia and Apple documentation conflict I would believe the latter.

In particular, it's not completely beyond belief that Apple might have a custom CSR format and process. (It would, arguably, be very Apple of them to do so.) If this involves a more secure method of submitting keys to Apple than plaintext, that would (initially appear to) avoid 3.

Fortunately, this is not only not so, but also easily mostly verifiably not so. Amending the previous suggestion:

> The CSR file contains the only the public key and a signed hash of the request itself; this can be confirmed by extracting its contents with OpenSSL:
>
> `openssl asn1parse -i -in CertificateSigningRequest.certSigningRequest`
> `openssl req -in CertificateSigningRequest.certSigningRequest -text -noout`

(I suppose someone might believe that the signature actually encodes the private key somehow; but demonstrating otherwise is well beyond the scope of this parenthetical, and also not too difficult for the interested paranoid to confirm without further explanation than this.)

The above lines may also serve as a reference for us: we can use these commands to peek at the inside of any received CSRs, to confirm that they are what they should be and nothing more.

Reply:
Nice, thanks, I'll use this!

I was hoping we could delight readers' ears with the phrase "contradictory doc", but this is a better solution. 😝

Reply:
> In particular, it's not completely beyond belief that Apple might have a custom CSR format and process. (It would, arguably, be very Apple of them to do so.)

It could be very Apple of them to have something custom -- but it'd be very un-Apple of them to have a system that involved them touching everyone's secret keys, when it's cryptographically possible to avoid that.

Helping the reader inspect things for themself, with these `openssl` commands, is a good solution.